### ISSUE TITLE
UI is super slow when the history contains nested lists with more than 2 levels

### ISSUE THREAD

**Issue by lldelisle** [galaxyproject/galaxy#16878]
**Describe the bug**
If you have a nested list of dataset collections in your history, the UI is super slow

**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 23.1
usegalaxy.org usegalaxy.eu

**To Reproduce**
1. Create a fresh history
2. Generate 2 datasets (I created texts with a single line)
3. Create a collection with these 2 datasets
4. Create a super nested collection with 'Apply rule' on this collection with the following rule:
```json
{
  "rules": [
    {
      "type": "add_column_metadata",
      "value": "identifier0"
    },
    {
      "type": "add_column_regex",
      "target_column": 0,
      "expression": "(.*)",
      "replacement": "label_\\1",
      "group_count": null
    },
    {
      "type": "add_column_regex",
      "target_column": 0,
      "expression": "(.*)",
      "replacement": "sample\\1",
      "group_count": null
    }
  ],
  "mapping": [
    {
      "type": "list_identifiers",
      "columns": [
        2,
        1,
        0
      ],
      "editing": false
    }
  ]
}
```
5. Use for example Apply Rule or try to run a workflow from this history.

**Expected behavior**
I expected a regular speed for this history, however, displaying things into the middle panel is super slow...

**Additional context**
This has also been experienced by @wm75 .

**Comment by mvdbeek** [comment]
This is the bug report of the year, sharp observation about something we were vaguely circling around already, and then you provide steps with exactly what's needed to reproduce, thanks so much @wm75 and @lldelisle

**Comment by lldelisle** [comment]
This means that there is no limit and no downside for super nesting or still the less is the better?

**Comment by lldelisle** [comment]
This is deployed on EU so I could test. Most of tools are much more responsive than before but "Apply Rules" is still super super slow (about 1 minute).

**Comment by lldelisle** [comment]
![Peek 2023-10-18 23-12](https://github.com/galaxyproject/galaxy/assets/30404086/c890721b-3d23-4338-acfd-b499384b8bb1)

**Comment by mvdbeek** [comment]
So after spending most of today on this and trying a lot of different optimizations, I noticed that

```
SELECT
    hda.extension AS extension
FROM
    history_dataset_association AS hda
WHERE
    hda.id IN (
        SELECT dce3.hda_id
        FROM dataset_collection_element AS dce1
        JOIN dataset_collection_element AS dce2 ON dce1.child_collection_id = dce2.dataset_collection_id
        JOIN dataset_collection_element AS dce3 ON dce2.child_collection_id = dce3.dataset_collection_id
        WHERE dce1.dataset_collection_id = 5523001
    )
```

is slow and

```
SELECT
    hda.extension AS extension
FROM
    history_dataset_association AS hda
WHERE
    hda.id IN (
        SELECT dce3.hda_id
        FROM dataset_collection_element AS dce1
        JOIN dataset_collection_element AS dce2 ON dce1.child_collection_id = dce2.dataset_collection_id
        JOIN dataset_collection_element AS dce3 ON dce2.child_collection_id = dce3.dataset_collection_id
        WHERE dce1.dataset_collection_id = 5523001 OFFSET 0
```
is fast. Taking out the join on the HDAs is also fast. The OFFSET 0 prevents the folding of the subquery into the main query by the query planner. Since the query planner was doing such a bad job @natefoo ran ANALYZE on the dataset_collection and dataset_collection_elements table and that improved the unmodified query again by a factor of about 1000 and turns the `dataset_collection_element AS dce3 ON dce2.child_collection_id = dce3.dataset_collection_id` scan from a seq scan into an index scan .... things seems pretty fast on usegalaxy.org now, maybe this also needs to be done on usegalaxy.eu ?

**Comment by bgruening** [comment]
I just did a:

```
ANALYZE  dataset_collection_element;
ANALYZE dataset_collection;

**Comment by lldelisle** [comment]
I don't understand. On org, now it is quick but on eu it is still slow.

**Comment by lldelisle** [comment]
![Peek 2023-10-19 22-07](https://github.com/galaxyproject/galaxy/assets/30404086/5e1d846b-bc61-48c2-afcd-28c980c3aa6b)

**Comment by lldelisle** [comment]
(Both histories and collections were created yesterday so before both of you did the ANALYSE)

**Comment by mvdbeek** [comment]
Yes, we also noticed this and looking into that with @bgruening. But I think it's safe to say

> This means that there is no limit and no downside for super nesting or still the less is the better?

that there is no fundamental problem with deeper nesting ... we'll get to them bottom of this eventually.

**Comment by lldelisle** [comment]
OK thank you very much for spending so much time on it...

**Comment by lldelisle** [comment]
Ping me if you need help.
